calc/vxlan_resolver.go

```diff
@@ -0,0 +1,453 @@
+			// remove it from internal state.
+			deletes.Add(*r)
+			delete(c.blockToRoutes[key], r.Key())
+		}
+
+		// At this point we've determined the correct diff to perform based on the block update.
+		// Delete any routes which are gone for good, withdraw modified routes, and send updates for
+		// new ones.
+		c.deleteRoutes(deletes)
+		c.withdrawRoutes(moved)
+		c.sendRouteUpdates(updates)
+	} else {
+		// Block has been deleted. Clean up routes that were contributed by this block.
+		routes := c.blockToRoutes[key]
+		c.deleteRoutes(set.From(routes))
+		delete(c.blockToRoutes, key)
+	}
+	return
+}
+
+// OnHostIPUpdate gets called whenever a node IP address changes. On an add/update,
+// we need to check if there are VTEPs or routes which are now valid, and trigger programming
+// of them to the data plane. On a delete, we need to withdraw any routes and VTEPs associated
+// with the node.
+func (c *VXLANResolver) OnHostIPUpdate(update api.Update) (_ bool) {
+	nodeName := update.Key.(model.HostIPKey).Hostname
+	logCxt := logrus.WithField("node", nodeName)
+	if update.Value != nil {
+		// Host IP updated or added. If it was added, we should check to see if we're ready
+		// to send a VTEP and associated routes. If we already knew about this one, we need to
+		// see if it has changed. If it has, we should remove and reprogram the VTEP and routes.
+		newIP := update.Value.(*net.IP).String()
+		currIP := c.nodeNameToIPAddr[nodeName]
+		c.nodeNameToIPAddr[nodeName] = newIP
+		logCxt = logCxt.WithFields(logrus.Fields{"newIP": newIP, "currIP": currIP})
+		if c.sentVTEPs[nodeName] {
+			if currIP == newIP {
+				// If we've already handled this node, there's nothing to do. Deduplicate.
+				logCxt.Debug("Skipping duplicate node IP update")
+				return
+			}
+
+			// We've already sent a VTEP for this node, and the node's IP address has changed.
+			// We need to revoke it and any routes before sending any updates.
+			routes := c.nodeNameToSentRoutes[nodeName]
+			c.withdrawRoutes(routes)
+			c.sendVTEPRemove(nodeName)
+		}
+
+		// Try sending a VTEP update. If we do, this will trigger a kick of any
+		// pending routes which might now be ready to send.
+		c.sendVTEPUpdate(nodeName)
+	} else {
+		// Withdraw any routes which target this VTEP, followed by the VTEP itself.
+		delete(c.nodeNameToIPAddr, nodeName)
+		routes := c.nodeNameToSentRoutes[nodeName]
+		c.withdrawRoutes(routes)
+		c.sendVTEPRemove(nodeName)
+	}
+	return
+}
+
+// OnHostConfigUpdate gets called whenever a node's host config changes. We only care about
+// VXLAN tunnel address updates. On an add/update, we need to check if there are VTEPs or routes which
+// are now valid, and trigger programming of them to the data plane. On a delete, we need to withdraw any
+// routes and VTEPs associated with the node.
+func (c *VXLANResolver) OnHostConfigUpdate(update api.Update) (_ bool) {
+	switch update.Key.(model.HostConfigKey).Name {
+	case "IPv4VXLANTunnelAddr":
```

Comment:
I can try to rework to use the Node object here - it shouldn't be too hard. And, I agree I like that better. 

One potential side effect of that is that it will no longer be a felixconfig, so won't be configurable via environment variable, etc. This is probably desirable for this though - I think we want it to come through the node.

Reply:
I think we should do this as a follow on - I'll raise something to track.